### PR TITLE
Support running test flows on WSL

### PR DESCRIPTION
**Pull request by vitalybu**
The purpose of this PR is to add support for running the test flows on WSL.
Docker for Windows is currently unable to forward traffic from the host to Linux based containers.
To workaround this limitation, when running on WSL, we expose the UCC ports to the host machine.
Once the ports are exposed, the `test_flows.py` script opens the CAPI sockets to the localhost and is able to communicate with the listeners in the containers.

### CODE REVIEW

File: tools/docker/run.sh [hunk 1]
```diff
@@ -107,7 +110,7 @@ main() {
                 -e INSTALL_DIR=${installdir}
                 --privileged
                 --network ${NETWORK}
-                --expose ${PORT}
+                --expose ${PORT} ${PUBLISH}
```
Comment (tomereli) [inline tools/docker/run.sh]: Any reason why not to always publish all ports?
What is the difference between exposing port and publishing it?

**Review by tomereli — APPROVED** [review]
@arnout or @rmelotte should also review this.

File: tools/docker/run.sh (hunk 1 above)
Comment (vitalybu) [reply tools/docker/run.sh]: You don't always want to publish the ports to the host.
Sometimes you only want to expose certain ports for container-to-container communication.

From the official docker documentation:

> The EXPOSE instruction defines the initial incoming ports that provide services. These ports are available to processes inside the container. An operator can use the --expose option to add to the exposed ports.
> 
> To expose a container’s internal port, an operator can start the container with the -P or -p flag. The exposed port is accessible on the host and the ports are available to any client that can reach the host.

File: tests/test_flows.py [hunk 2]
```diff
@@ -223,6 +234,14 @@ def init(self):
     def _check_log_internal(self, device: str, program: str, regex: str) -> bool:
         '''Search for regex in logfile for program on device.'''
         logfilename = os.path.join(self.rootdir, 'logs', device, 'beerocks_{}.log'.format(program))
+
+        # WSL doesn't support symlinks on NTFS, so resolve the symlink manually
```
Comment (rmelotte) [inline tests/test_flows.py:239]: What does WSL use NTFS for? (I know nothing about WSL)

File: tools/docker/runner/Dockerfile [hunk 3]
```diff
@@ -39,4 +39,4 @@ ADD start-controller /usr/bin
 ADD start-agent /usr/bin
 ADD start-controller-agent /usr/bin
 
-ENTRYPOINT ["/root/entrypoint.sh"]
+ENTRYPOINT ["/bin/bash", "/root/entrypoint.sh"]
```
Comment (rmelotte) [inline tools/docker/runner/Dockerfile]: Why do you want to explicitly call bash instead of relying on the shebang?

File: tools/docker/runner/entrypoint.sh [hunk 4]
```diff
@@ -14,6 +14,12 @@ run() {
 bridge_ip="$1"; shift
 base_mac="$1"; shift
 
+# If the requested bridge ip is 0.0.0.0, use the ip address that
+# was allocated by the daemon to this container (from eth0)
+if [[ ${bridge_ip} == 0.0.0.0* ]]; then
+    bridge_ip="$(ip addr show dev eth0 | grep inet | awk '{print $2}')"
+fi
```
Comment (rmelotte) [inline tools/docker/runner/entrypoint.sh:21]: I proposed to do this a while ago, so if you feel like it this could be done even for linux : https://github.com/prplfoundation/prplMesh/issues/461

File: tools/docker/run.sh [hunk 5]
```diff
@@ -64,7 +65,8 @@ main() {
             -n | --name)        NAME="$2"; shift; shift ;;
             -u | --unique-id)   UNIQUE_ID="$2"; shift; shift ;;
             -t | --tag)         TAG=":$2"; shift; shift ;;
-            -p | --port)        PORT=":$2"; shift; shift ;;
+            -p | --port)        PORT="$2"; shift; shift ;;
+            -P | --publish)     PUBLISH="-P"; shift ;;
```
Comment (rmelotte) [inline tools/docker/run.sh]: From `man docker-run`:
>       -P, --publish-all=true|false
>       Publish all exposed ports to random ports on the host interfaces. The default is false.

Are you sure you want to expose them to random ports on the host?

If not you could use this syntax to control what the ports are mapped to:
>       -p, --publish ip:[hostPort]:containerPort | [hostPort:]containerPort
>          Publish a container's port, or range of ports, to the host.

**More importantly**: I don't think you need them to be published externally, do you?
If not, publish them to "localhost" on the host instead of "0.0.0.0" (which is what "-P" does, at least under Linux).

File: tests/test_gw_repeater.sh [hunk 6]
```diff
@@ -27,6 +27,35 @@ usage() {
...
+check_wsl() {
+    # Do nothing for non-WSL environments
+    if ! grep -q Microsoft /proc/version; then return; fi
+
+    status "Running in WSL"
+    
+    # Read GW & Repeater UCC ports
+    local GW_UCC_PORT=$(grep ucc_listener_port \
+    ${rootdir}/build/install/config/beerocks_controller.conf | \
+    awk -F'[= ]' '{ print $2 }')
```
Comment (rmelotte) [inline tests/test_gw_repeater.sh]: I know there are already many shellcheck issues in this script, but let's not introduce more
```suggestion
    local GW_UCC_PORT
    GW_UCC_PORT="$(grep ucc_listener_port \
    "${rootdir}/build/install/config/beerocks_controller.conf" | \
    awk -F'[= ]' '{ print $2 }')"
```

File: tests/test_gw_repeater.sh [hunk 7]
```diff
@@ -27,6 +27,35 @@ usage() {
...
+
+    local RP_UCC_PORT=$(grep ucc_listener_port \
+    ${rootdir}/build/install/config/beerocks_agent.conf | \
+    awk -F'[= ]' '{ print $2 }')
```
Comment (rmelotte) [inline tests/test_gw_repeater.sh:51]: Same here

**Review by rmelotte — APPROVED** [review]
Not requesting changes to not block you, but I think you'll want to at least address the comment on using `-P` with `docker run`.

File: tests/test_flows.py (hunk 2 above)
Comment (vitalybu) [reply tests/test_flows.py:239]: Actually it's a combination of issues. WSL doesn't use NTFS directly, it does implement a somewhat Linux compatible filesystem.
The problem is with Docker for Windows. There's a known bug that prevents the daemon (that actually runs on a Hyper-V VM) from mounting non-NTFS volumes into the containers.
Since we map the `build` and `tmp` folders into the `prplmesh-runner`, all the logs are eventually created on an NTFS filesystem that doesn't support symlinks.

File: tools/docker/runner/Dockerfile (hunk 3 above)
Comment (vitalybu) [reply tools/docker/runner/Dockerfile]: Not sure actually :)
I'll revert and update the `entrypoint.sh` to `#!/bin/bash`

File: tools/docker/runner/entrypoint.sh (hunk 4 above)
Comment (vitalybu) [reply tools/docker/runner/entrypoint.sh:21]: Maybe we can skip generating IP addresses at all.
Not sure why we need if the daemon does it anyway.

File: tools/docker/run.sh (hunk 5 above)
Comment (vitalybu) [reply tools/docker/run.sh]: Publish just makes these ports accessible to the host.
I do prefer random ports (which we can later read by `docker port`) instead of managing them myself.

File: tests/test_flows.py (hunk 2 above)
Comment (rmelotte) [reply tests/test_flows.py:239]: Ok, thanks for the explanation :-)

File: tools/docker/runner/entrypoint.sh (hunk 4 above)
Comment (rmelotte) [reply tools/docker/runner/entrypoint.sh:21]: Yes exactly.

File: tools/docker/run.sh (hunk 5 above)
Comment (rmelotte) [reply tools/docker/run.sh]: `--publish` (which is not the option you used) will make it accept connections from whatever IP you specified.
`--publish-all` will make it accept connections from anywhere.

Still from `man docker-run`:
>When set to true publish all exposed ports to the host interfaces. The default is false. If the operator uses -P (or -p) then Docker will make the exposed port accessible on the host and
       **the ports will be available to any client that can reach the host**. When using -P, Docker will bind any exposed port to a random port on the host within an ephemeral port range defined by
>       /proc/sys/net/ipv4/ip_local_port_range. To find the mapping between the host ports and the exposed ports, use docker port(1).

In other words, you _have_ to use `--publish 127.0.0.1::<capi_port>` unless you want the CAPI listener to be available to anyone on your local network.

(you will still have random ports on the host if you specify the IP and the container port only)

File: tests/test_flows.py (hunk 2 above)
Comment (ghost) [reply tests/test_flows.py:239]: Actually, NTFS does support symlinks (since NTFS version 3.0, which was released together with Windows 2000). Windows API support (e.g. OpenFile) for symlinks was added to Windows in Windows Vista, though linux (using ntfs-3g) could create these symlinks on NTFS even before then.
You can create those symlinks on Windows with the `mklink` command.
Annoyingly, IIRC you always need administrator privileges to create symlinks on windows though.

Comment (ghost) [reply tests/test_flows.py:239]: Actually, it seems my information is outdated: https://blogs.windows.com/windowsdeveloper/2016/12/02/symlinks-windows-10/

Comment (vitalybu) [reply tests/test_flows.py:239]: Symlinks work fine when created from Windows using `mklink`, but when created in WSL, they appear as regular files on NTFS.

File: tools/docker/run.sh (hunk 5 above)
Comment (vitalybu) [reply tools/docker/run.sh]: Updated the script to allow publishing ports on the localhost only.

File: tests/test_gw_repeater.sh (hunk 6 above)
Comment (vitalybu) [reply tests/test_gw_repeater.sh]: Done. Also fixed all shellcheck issues while I'm at it.
